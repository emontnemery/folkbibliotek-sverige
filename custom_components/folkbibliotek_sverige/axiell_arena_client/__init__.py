"""Python client for Axiell Arena."""

from __future__ import annotations

from dataclasses import dataclass
import logging
from typing import TYPE_CHECKING, Self

from bs4 import BeautifulSoup, NavigableString, Tag

if TYPE_CHECKING:
    import aiohttp

_LOGGER = logging.getLogger(__name__)

LOGIN_ATTEMPTS = 3


class ArenaError(Exception):
    """Base class for exceptions in this module."""


class ArenaLoginError(ArenaError):
    """Exception raised for errors in the login process."""

    msg = "Could not log in"

    def __init__(self) -> None:
        """Initialize the exception."""
        super().__init__(self.msg)


class ArenaLoginFormNotFoundError(ArenaLoginError):
    """Exception raised when the login form is not found."""

    msg = "Could not find login form"


class ArenaAccountLockedError(ArenaLoginError):
    """Exception raised when the account is locked."""

    msg = "Account is locked"


class ArenaInvalidCredentialsError(ArenaLoginError):
    """Exception raised when the credentials are not valid."""

    msg = "Invalid credentials"


class ArenaNotLoggedInError(ArenaError):
    """Exception raised when not logged in."""

    msg = "Not logged in"


def _get_value(tag: Tag, selector: str) -> str | None:
    selected = tag.select_one(f"{selector}  span.arena-value")
    if not selected:
        return None
    return selected.text


@dataclass
class LibraryMaterial:
    """Library material container."""

    record_id: str | None = None
    type: str | None = None
    title: str | None = None
    author: str | None = None
    year: str | None = None


@dataclass
class LibraryLoan(LibraryMaterial):
    """Library loan container."""

    loan_date: str | None = None
    expire_date: str | None = None
    renewable: bool | None = None

    @classmethod
    def from_soup(cls, soup: Tag) -> Self:
        """Construct a library loan from a BeautifulSoup tag."""
        renewable = False
        if "arena-renewal-false" in soup.get("class"):
            renewable = False
        elif "arena-renewal-true" in soup.get("class"):
            renewable = True
        return cls(
            record_id=soup.select_one("span.arena-record-id").text,
            type=_get_value(soup, "div.arena-record-media"),
            title=soup.select_one("div.arena-record-title span").text,
            author=_get_value(soup, "div.arena-record-author"),
            year=_get_value(soup, "div.arena-record-year"),
            loan_date=_get_value(soup, "div.arena-renewal-branch").rsplit(" ", 1)[-1],
            expire_date=soup.select_one("span.arena-renewal-date-value").text,
            renewable=renewable,
        )


@dataclass
class LibraryReservation(LibraryMaterial):
    """Library reservation container."""

    created_date: str | None = None
    expire_date: str | None = None
    queue_number: str | None = None
    pickup_library: str | None = None

    @classmethod
    def from_soup(cls, soup: Tag) -> Self:
        """Construct a library reservation from a BeautifulSoup tag."""
        return cls(
            record_id=soup.select_one("span.arena-record-id").text,
            type=_get_value(soup, "div.arena-record-media"),
            title=soup.select_one("div.arena-record-title span").text,
            author=_get_value(soup, "div.arena-record-author"),
            year=_get_value(soup, "div.arena-record-year"),
            created_date=_get_value(soup, "td.arena-reservation-from-container"),
            expire_date=_get_value(soup, "td.arena-reservation-to-container"),
            queue_number=_get_value(soup, "td.arena-record-queue").split(" ")[0],
            pickup_library=_get_value(soup, "td.arena-record-branch"),
        )


@dataclass
class LibraryReservationReady(LibraryMaterial):
    """Library reservation ready container."""

    created_date: str | None = None
    pickup_date: str | None = None
    reservation_number: str | None = None
    pickup_library: str | None = None

    @classmethod
    def from_soup(cls, soup: Tag) -> Self:
        """Construct a library reservation from a BeautifulSoup tag."""
        return cls(
            record_id=soup.select_one("span.arena-record-id").text,
            type=_get_value(soup, "div.arena-record-media"),
            title=soup.select_one("div.arena-record-title span").text,
            author=_get_value(soup, "div.arena-record-author"),
            year=_get_value(soup, "div.arena-record-year"),
            created_date=_get_value(soup, "td.arena-reservation-from-container"),
            pickup_date=_get_value(soup, "td.arena-record-expire"),
            reservation_number=_get_value(soup, "td.arena-record-pickup"),
            pickup_library=_get_value(soup, "td.arena-record-branch"),
        )


@dataclass
class LibraryDebt(LibraryMaterial):
    """Library debt container."""

    fee_date: str | None = None
    fee_type: str | None = None
    fee_amount: str | None = None


class ArenaClient:
    """Axiell Arena client."""

    def __init__(
        self, *, session: aiohttp.ClientSession, url: str, username: str, password: str
    ) -> None:
        """Initialize the client."""
        self.session = session
        self.base_url = url
        self.username = username
        self.password = password

    async def get_account_overview_html(self) -> BeautifulSoup:
        """Get the account overview."""
        url = f"{self.base_url}/protected/my-account/overview"
        return await self._get_url(url)

    async def get_account_overview(self) -> BeautifulSoup:
        """Get the account overview and wrap it in a BeautifulSoup object."""
        return BeautifulSoup(await self.get_account_overview_html(), "html.parser")

    def get_loans(self, soup: BeautifulSoup) -> list[LibraryLoan]:
        """Get library loans."""
        table: Tag = soup.select_one("table#loansTable")
        if not table:
            return []
        _LOGGER.debug("Loans: %s", table)
        return [
            LibraryLoan.from_soup(row)
            for row in table.children
            if not isinstance(row, NavigableString) and row.name == "tr"
        ]

    def _get_reservations(self, soup: BeautifulSoup) -> Tag | None:
        reservations = soup.select_one("div.portlet-myReservations")
        if not reservations or not reservations.select("div.arena-library-record"):
            _LOGGER.debug("No reservations")
            return None
        _LOGGER.debug("Reservations: %s", reservations)
        return reservations

    def get_active_reservations(self, soup: BeautifulSoup) -> list[LibraryReservation]:
        """Get library reservations."""
        if not (reservations := self._get_reservations(soup)):
            return []
        return [
            LibraryReservation.from_soup(reservation)
            for reservation in reservations.select("div.arena-library-record")
            if not reservation.select_one("td.arena-record-pickup")
        ]

    def get_ready_reservations(
        self, soup: BeautifulSoup
    ) -> list[LibraryReservationReady]:
        """Get library reservations."""
        if not (reservations := self._get_reservations(soup)):
            return []
        return [
            LibraryReservationReady.from_soup(reservation)
            for reservation in reservations.select("div.arena-library-record")
            if reservation.select_one("td.arena-record-pickup")
        ]

    async def _get_url(self, url: str) -> str:
        try:
            async with self.session.get(url) as resp:
                resp.raise_for_status()
                text = await resp.text()
                self._raise_if_not_logged_in(text)
                return text
        except ArenaNotLoggedInError:
            return await self._login_and_get_url(url)

    async def _login_and_get_url(self, url: str) -> str:
        query_params = {
            "p_p_id": "patronLogin_WAR_arenaportlet",
            "p_p_lifecycle": 1,
            "p_p_state": "normal",
            "p_p_mode": "view",
            "_patronLogin_WAR_arenaportlet__wu": (
                "/patronLogin/?wicket:interface=:1:signInPanel:signInFormPanel"
                ":signInForm::IFormSubmitListener::"
            ),
        }
        form_data = {
            "id__patronLogin__WAR__arenaportlet____3_hf_0": "",
            "openTextUsernameContainer:openTextUsername": self.username,
            "textPassword": self.password,
        }
        self.session.cookie_jar.clear()

        attempt = 0
        while attempt < LOGIN_ATTEMPTS:
            attempt += 1
            _LOGGER.debug("Attempting to log in, attempt %d", attempt)
            async with self.session.post(
                url, params=query_params, data=form_data
            ) as resp:
                resp.raise_for_status()
                text = await resp.text()
                try:
                    self._raise_if_not_logged_in(text)
                except ArenaNotLoggedInError:
                    continue
                return text
        raise ArenaLoginError("Unknown login error")  # noqa: TRY003, EM101

    def _raise_if_not_logged_in(self, text: str) -> None:
        soup = BeautifulSoup(text, "html.parser")
        if not (
            patron_login := soup.find(
                "div", {"id": "portlet_patronLogin_WAR_arenaportlet"}
            )
        ):
            raise ArenaLoginFormNotFoundError
        if not (patron_login.find("div", {"class": "arena-patron-form"})):
            _LOGGER.debug("Logged in")
            return
        if not (
            feedback_panel := patron_login.find(
                "span", {"class": "feedbackPanelWARNING"}
            )
        ):
            _LOGGER.debug("No feedback panel found: %s", text)
            raise ArenaNotLoggedInError("No feedback panel found: %s", text)  # noqa: TRY003, EM101
        if feedback_panel.text.startswith("Ditt konto har st√§ngts"):
            _LOGGER.debug("Account locked")
            raise ArenaAccountLockedError
        if feedback_panel.text.startswith("Du blev inte inloggad"):
            _LOGGER.debug("Invalid credentials")
            raise ArenaInvalidCredentialsError
        _LOGGER.debug("Unknown error while logging in: %s", feedback_panel.text)
        raise ArenaNotLoggedInError(feedback_panel.text)
