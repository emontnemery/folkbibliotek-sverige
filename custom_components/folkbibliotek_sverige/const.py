"""Constants for the Folkbibliotek Sverige integration."""

import logging

DOMAIN = "folkbibliotek_sverige"

LOGGER = logging.getLogger(__name__)
