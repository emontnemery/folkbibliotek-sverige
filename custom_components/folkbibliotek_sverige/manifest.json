{
  "domain": "folkbibliotek_sverige",
  "name": "Folkbibliotek Sverige",
  "codeowners": [
    "@emontnemery"
  ],
  "config_flow": true,
  "documentation": "https://github.com/emontnemery/folkbibliotek-sverige",
  "integration_type": "service",
  "iot_class": "cloud_polling",
  "issue_tracker": "https://github.com/emontnemery/folkbibliotek-sverige/issues",
  "requirements": ["beautifulsoup4>=4.13.3"],
  "version": "0.0.1"
}