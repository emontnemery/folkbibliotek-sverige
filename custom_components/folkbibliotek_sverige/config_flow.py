"""Config flow for the Folkbibliotek Sverige integration."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from homeassistant.config_entries import ConfigFlow, ConfigFlowResult
from homeassistant.const import CONF_NAME, CONF_PASSWORD, CONF_URL, CONF_USERNAME
from homeassistant.helpers import selector
from homeassistant.helpers.aiohttp_client import async_get_clientsession
import voluptuous as vol

from .axiell_arena_client import (
    ArenaAccountLockedError,
    ArenaClient,
    ArenaError,
    ArenaInvalidCredentialsError,
)
from .const import DOMAIN, LOGGER

if TYPE_CHECKING:
    from collections.abc import Mapping

    from homeassistant.core import HomeAssistant

STEP_REAUTH_DATA_SCHEMA = vol.Schema(
    {
        vol.Required(CONF_URL): selector.TextSelector(
            selector.TextSelectorConfig(read_only=True)
        ),
        vol.Required(CONF_USERNAME): selector.TextSelector(
            selector.TextSelectorConfig(read_only=True)
        ),
        vol.Required(CONF_PASSWORD): str,
    }
)

STEP_USER_DATA_SCHEMA = vol.Schema(
    {
        vol.Required(CONF_NAME): str,
        vol.Required(CONF_URL): str,
        vol.Required(CONF_USERNAME): str,
        vol.Required(CONF_PASSWORD): str,
    }
)


async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> None:
    """
    Validate the user input allows us to connect.

    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
    """
    client = ArenaClient(
        session=async_get_clientsession(hass),
        url=data[CONF_URL],
        username=data[CONF_USERNAME],
        password=data[CONF_PASSWORD],
    )

    await client.get_account_overview()


class FolkbibliotekSverigeConfigFlow(ConfigFlow, domain=DOMAIN):
    """Handle a config flow for Folkbibliotek Sverige."""

    VERSION = 1

    async def async_step_user(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Handle the initial step."""
        errors: dict[str, str] = {}
        if user_input is not None:
            self._async_abort_entries_match(
                {
                    CONF_URL: user_input[CONF_URL],
                    CONF_USERNAME: user_input[CONF_USERNAME],
                }
            )
            try:
                await validate_input(self.hass, user_input)
            except ArenaError:
                errors["base"] = "cannot_connect"
            except ArenaAccountLockedError:
                errors["base"] = "account_locked"
            except ArenaInvalidCredentialsError:
                errors["base"] = "invalid_credentials"
            except Exception:  # noqa: BLE001
                LOGGER.exception("Unexpected exception")
                errors["base"] = "unknown"
            else:
                return self.async_create_entry(
                    title=user_input[CONF_NAME], data=user_input
                )

        return self.async_show_form(
            step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
        )

    async def async_step_reauth(
        self,
        entry_data: Mapping[str, Any],  # noqa: ARG002
    ) -> ConfigFlowResult:
        """Handle configuration by re-auth."""
        return await self.async_step_reauth_confirm()

    async def async_step_reauth_confirm(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Handle the initial step."""
        errors: dict[str, str] = {}
        config_entry_data = self._get_reauth_entry().data

        if user_input is not None:
            try:
                await validate_input(self.hass, config_entry_data | user_input)
            except ArenaError:
                errors["base"] = "cannot_connect"
            except ArenaAccountLockedError:
                errors["base"] = "account_locked"
            except ArenaInvalidCredentialsError:
                errors["base"] = "invalid_credentials"
            except Exception:  # noqa: BLE001
                LOGGER.exception("Unexpected exception")
                errors["base"] = "unknown"
            else:
                return self.async_update_reload_and_abort(
                    self._get_reauth_entry(), data_updates=user_input
                )

        return self.async_show_form(
            step_id="reauth_confirm",
            data_schema=self.add_suggested_values_to_schema(
                STEP_REAUTH_DATA_SCHEMA,
                {
                    CONF_URL: config_entry_data[CONF_URL],
                    CONF_USERNAME: config_entry_data[CONF_USERNAME],
                },
            ),
            errors=errors,
        )
