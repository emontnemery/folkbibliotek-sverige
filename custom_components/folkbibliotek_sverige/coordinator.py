"""DataUpdateCoordinator for the Folkbibliotek Sverige integration."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import timedelta
from typing import TYPE_CHECKING

from homeassistant.exceptions import ConfigEntryAuthFailed
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed

from .axiell_arena_client import (
    ArenaAccountLockedError,
    ArenaClient,
    ArenaError,
    ArenaInvalidCredentialsError,
    LibraryLoan,
    LibraryReservation,
    LibraryReservationReady,
)
from .const import DOMAIN, LOGGER

if TYPE_CHECKING:
    from homeassistant.core import HomeAssistant

    from . import FolkbibliotekSverigeConfigEntry


@dataclass(frozen=True, kw_only=True)
class FolkbibliotekSverigeData:
    """Analytics data class."""

    loans: list[LibraryLoan]
    active_reservations: list[LibraryReservation]
    waiting_reservations: list[
        LibraryReservationReady
    ]  # Rename to LibraryWaitingReservation


class FolkbibliotekSverigeDataUpdateCoordinator(
    DataUpdateCoordinator[FolkbibliotekSverigeData]
):
    """Folkbibliotek Sverige data update coordinator."""

    config_entry: FolkbibliotekSverigeConfigEntry

    def __init__(
        self,
        hass: HomeAssistant,
        config_entry: FolkbibliotekSverigeConfigEntry,
        client: ArenaClient,
    ) -> None:
        """Initialize the data coordinator."""
        super().__init__(
            hass,
            LOGGER,
            config_entry=config_entry,
            name=DOMAIN,
            update_interval=timedelta(hours=2),
        )
        self._client = client

    async def _async_update_data(self) -> FolkbibliotekSverigeData:
        if isinstance(self.last_exception, ConfigEntryAuthFailed):
            # If the last exception was a ConfigEntryAuthFailed, we should not try
            # to update again, or the account will be locked.
            LOGGER.warning(
                "Authentication failed, not trying to update again. "
                "Please check your credentials and try again."
            )
            raise self.last_exception
        try:
            overview = await self._client.get_account_overview()
        except (ArenaAccountLockedError, ArenaInvalidCredentialsError) as err:
            raise ConfigEntryAuthFailed(err) from err
        except ArenaError as err:
            raise UpdateFailed("Error communicating with the library database") from err  # noqa: EM101, TRY003
        LOGGER.debug("loans: %s", self._client.get_loans(overview))
        LOGGER.debug(
            "active reservations: %s", self._client.get_active_reservations(overview)
        )
        LOGGER.debug(
            "ready reservations: %s", self._client.get_ready_reservations(overview)
        )
        return FolkbibliotekSverigeData(
            loans=self._client.get_loans(overview),
            active_reservations=self._client.get_active_reservations(overview),
            waiting_reservations=self._client.get_ready_reservations(overview),
        )
