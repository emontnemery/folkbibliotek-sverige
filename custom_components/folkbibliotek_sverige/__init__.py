"""
The Folkbibliotek Sverige custom integration.

For more details about this integration, please refer to
https://github.com/emontnemery/folkbibliotek-sverige
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import Platform
from homeassistant.helpers.aiohttp_client import async_get_clientsession

from .axiell_arena_client import ArenaClient
from .coordinator import FolkbibliotekSverigeDataUpdateCoordinator

if TYPE_CHECKING:
    from homeassistant.core import HomeAssistant


_PLATFORMS: list[Platform] = [Platform.TODO]

type FolkbibliotekSverigeConfigEntry = ConfigEntry[FolkbibliotekSverigeData]


@dataclass(frozen=True)
class FolkbibliotekSverigeData:
    """FolkbibliotekSverige data class."""

    coordinator: FolkbibliotekSverigeDataUpdateCoordinator


async def async_setup_entry(
    hass: HomeAssistant, entry: FolkbibliotekSverigeConfigEntry
) -> bool:
    """Set up Folkbibliotek Sverige from a config entry."""
    client = ArenaClient(
        session=async_get_clientsession(hass),
        url=entry.data["url"],
        username=entry.data["username"],
        password=entry.data["password"],
    )

    coordinator = FolkbibliotekSverigeDataUpdateCoordinator(hass, entry, client)
    await coordinator.async_config_entry_first_refresh()
    entry.runtime_data = FolkbibliotekSverigeData(coordinator=coordinator)

    entry.async_on_unload(entry.add_update_listener(update_listener))
    await hass.config_entries.async_forward_entry_setups(entry, _PLATFORMS)

    return True


async def async_unload_entry(
    hass: HomeAssistant, entry: FolkbibliotekSverigeConfigEntry
) -> bool:
    """Unload a config entry."""
    return await hass.config_entries.async_unload_platforms(entry, _PLATFORMS)


async def update_listener(
    hass: HomeAssistant, entry: FolkbibliotekSverigeConfigEntry
) -> None:
    """Handle options update."""
    await hass.config_entries.async_reload(entry.entry_id)
