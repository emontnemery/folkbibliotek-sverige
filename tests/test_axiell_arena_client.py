"""Test the Axiell Arena client."""

from __future__ import annotations

import re
from typing import TYPE_CHECKING

import aiohttp
from aioresponses import aioresponses
import pytest

from custom_components.folkbibliotek_sverige.axiell_arena_client import (
    ArenaAccountLockedError,
    ArenaClient,
    ArenaInvalidCredentialsError,
    ArenaLoginError,
)

from . import load_fixture

if TYPE_CHECKING:
    from collections.abc import AsyncGenerator, Generator

    from syrupy import SnapshotAssertion

BASE_URL = "https://folkbiblioteken.lund.se"
USERNAME = "username"
PASSWORD = "password"


@pytest.fixture
async def client() -> AsyncGenerator[ArenaClient]:
    """Return a go2rtc rest client."""
    async with (
        aiohttp.ClientSession() as session,
    ):
        client_ = ArenaClient(
            session=session, url=BASE_URL, username=USERNAME, password=PASSWORD
        )
        yield client_


@pytest.fixture(name="responses")
def aioresponses_fixture() -> Generator[aioresponses]:
    """Return aioresponses fixture."""
    with aioresponses() as mocked_responses:
        yield mocked_responses


async def test_success(
    responses: aioresponses,
    client: ArenaClient,
    snapshot: SnapshotAssertion,
) -> None:
    """Test successful retrieval of account overview."""
    responses.get(
        f"{BASE_URL}/protected/my-account/overview",
        body=load_fixture("logged_in.html"),
    )

    overview = await client.get_account_overview()
    assert client.get_loans(overview) == snapshot
    assert client.get_active_reservations(overview) == []
    assert client.get_ready_reservations(overview) == []


async def test_success_no_loans(
    responses: aioresponses,
    client: ArenaClient,
    snapshot: SnapshotAssertion,
) -> None:
    """Test successful retrieval of account overview with no checked out media."""
    responses.get(
        f"{BASE_URL}/protected/my-account/overview",
        body=load_fixture("logged_in_no_loans.html"),
    )

    overview = await client.get_account_overview()
    assert client.get_loans(overview) == []
    assert client.get_active_reservations(overview) == snapshot
    assert client.get_ready_reservations(overview) == []


async def test_success_no_reservations(
    responses: aioresponses,
    client: ArenaClient,
    snapshot: SnapshotAssertion,
) -> None:
    """Test successful retrieval of account overview with no holds."""
    responses.get(
        f"{BASE_URL}/protected/my-account/overview",
        body=load_fixture("logged_in_no_reservations.html"),
    )

    overview = await client.get_account_overview()
    assert client.get_loans(overview) == snapshot
    assert client.get_active_reservations(overview) == []
    assert client.get_ready_reservations(overview) == []


async def test_success_reservation_to_pick_up(
    responses: aioresponses,
    client: ArenaClient,
    snapshot: SnapshotAssertion,
) -> None:
    """Test successful retrieval of account overview with hold to pick up."""
    responses.get(
        f"{BASE_URL}/protected/my-account/overview",
        body=load_fixture("logged_in_reservation_to_pick_up.html"),
    )

    overview = await client.get_account_overview()
    assert client.get_loans(overview) == snapshot
    assert client.get_active_reservations(overview) == []
    assert client.get_ready_reservations(overview) == snapshot


async def test_success_need_login(
    responses: aioresponses,
    client: ArenaClient,
    snapshot: SnapshotAssertion,
) -> None:
    """Test log in."""
    responses.get(
        f"{BASE_URL}/protected/my-account/overview",
        body=load_fixture("not_logged_in.html"),
    )
    escaped_url = re.escape(f"{BASE_URL}/protected/my-account/overview")
    responses.post(
        re.compile(rf"^{escaped_url}\?_patronLogin_WAR.*"),
        body=load_fixture("logged_in.html"),
        repeat=True,
    )

    overview = await client.get_account_overview()
    assert client.get_loans(overview) == snapshot
    assert client.get_active_reservations(overview) == []
    assert client.get_ready_reservations(overview) == []


async def test_no_login(
    responses: aioresponses,
    client: ArenaClient,
) -> None:
    """Test no log in."""
    responses.get(
        f"{BASE_URL}/protected/my-account/overview",
        body=load_fixture("not_logged_in.html"),
    )
    escaped_url = re.escape(f"{BASE_URL}/protected/my-account/overview")
    responses.post(
        re.compile(rf"^{escaped_url}\?_patronLogin_WAR.*"),
        body=load_fixture("not_logged_in.html"),
        repeat=True,
    )

    with pytest.raises(ArenaLoginError):
        await client.get_account_overview()


async def test_account_locked(
    responses: aioresponses,
    client: ArenaClient,
) -> None:
    """Test account is locked."""
    responses.get(
        f"{BASE_URL}/protected/my-account/overview",
        body=load_fixture("not_logged_in.html"),
    )
    escaped_url = re.escape(f"{BASE_URL}/protected/my-account/overview")
    responses.post(
        re.compile(rf"^{escaped_url}\?_patronLogin_WAR.*"),
        body=load_fixture("login_failed_too_many_attempts.html"),
    )

    with pytest.raises(ArenaAccountLockedError):
        await client.get_account_overview()


async def test_wrong_credentials(
    responses: aioresponses,
    client: ArenaClient,
) -> None:
    """Test account wrong password."""
    responses.get(
        f"{BASE_URL}/protected/my-account/overview",
        body=load_fixture("not_logged_in.html"),
    )
    escaped_url = re.escape(f"{BASE_URL}/protected/my-account/overview")
    responses.post(
        re.compile(rf"^{escaped_url}\?_patronLogin_WAR.*"),
        body=load_fixture("login_failed_wrong_credentials.html"),
    )

    with pytest.raises(ArenaInvalidCredentialsError):
        await client.get_account_overview()
