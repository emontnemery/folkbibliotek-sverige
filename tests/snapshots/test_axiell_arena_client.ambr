# serializer version: 1
# name: test_success
  list([
    dict({
      'author': 'Maas, Sarah J',
      'expire_date': '2024-11-06',
      'loan_date': '2024-10-23',
      'record_id': '461553',
      'renewable': False,
      'title': 'Tower of dawn',
      'type': 'Bok',
      'year': '2017',
    }),
    dict({
      'author': 'Anderberg, Bengt',
      'expire_date': '2024-11-09',
      'loan_date': '2024-07-13',
      'record_id': '618578',
      'renewable': True,
      'title': 'Amorina',
      'type': 'Bok',
      'year': '2022',
    }),
    dict({
      'author': 'Maas, Sarah J',
      'expire_date': '2024-11-09',
      'loan_date': '2024-07-29',
      'record_id': '489758',
      'renewable': True,
      'title': 'Stormarnas imperium',
      'type': 'Bok',
      'year': '2019',
    }),
    dict({
      'author': 'Nowlin, Laura',
      'expire_date': '2024-11-09',
      'loan_date': '2024-09-28',
      'record_id': '599900',
      'renewable': False,
      'title': 'If he had been with me',
      'type': 'Bok',
      'year': '2019',
    }),
    dict({
      'author': 'Maas, Sarah J',
      'expire_date': '2024-11-19',
      'loan_date': '2024-10-22',
      'record_id': '497617',
      'renewable': False,
      'title': 'Kingdom of ash',
      'type': 'Bok',
      'year': '2018',
    }),
  ])
# ---
# name: test_success_need_login
  list([
    dict({
      'author': 'Maas, Sarah J',
      'expire_date': '2024-11-06',
      'loan_date': '2024-10-23',
      'record_id': '461553',
      'renewable': False,
      'title': 'Tower of dawn',
      'type': 'Bok',
      'year': '2017',
    }),
    dict({
      'author': 'Anderberg, Bengt',
      'expire_date': '2024-11-09',
      'loan_date': '2024-07-13',
      'record_id': '618578',
      'renewable': True,
      'title': 'Amorina',
      'type': 'Bok',
      'year': '2022',
    }),
    dict({
      'author': 'Maas, Sarah J',
      'expire_date': '2024-11-09',
      'loan_date': '2024-07-29',
      'record_id': '489758',
      'renewable': True,
      'title': 'Stormarnas imperium',
      'type': 'Bok',
      'year': '2019',
    }),
    dict({
      'author': 'Nowlin, Laura',
      'expire_date': '2024-11-09',
      'loan_date': '2024-09-28',
      'record_id': '599900',
      'renewable': False,
      'title': 'If he had been with me',
      'type': 'Bok',
      'year': '2019',
    }),
    dict({
      'author': 'Maas, Sarah J',
      'expire_date': '2024-11-19',
      'loan_date': '2024-10-22',
      'record_id': '497617',
      'renewable': False,
      'title': 'Kingdom of ash',
      'type': 'Bok',
      'year': '2018',
    }),
  ])
# ---
# name: test_success_no_loans
  list([
    dict({
      'author': 'Lindgren, Astrid',
      'created_date': '2024-11-09',
      'expire_date': '2026-11-09',
      'pickup_library': 'Stadsbiblioteket',
      'queue_number': '1',
      'record_id': '519954',
      'title': 'Do you know Pippi Longstocking?',
      'type': 'Bok',
      'year': '2019',
    }),
  ])
# ---
# name: test_success_no_reservations
  list([
    dict({
      'author': 'Simsion, Graeme',
      'expire_date': '2025-03-14',
      'loan_date': '2025-01-22',
      'record_id': '501688',
      'renewable': True,
      'title': 'The Rosie result',
      'type': 'Bok',
      'year': '2019',
    }),
    dict({
      'author': 'Cederlund, Charlotte',
      'expire_date': '2025-03-16',
      'loan_date': '2025-02-16',
      'record_id': '556230',
      'renewable': True,
      'title': 'Så jävla operfekt',
      'type': 'Bok',
      'year': '2021',
    }),
  ])
# ---
# name: test_success_reservation_to_pick_up
  list([
    dict({
      'author': 'Aveyard, Victoria',
      'expire_date': '2025-03-13',
      'loan_date': '2025-01-11',
      'record_id': '417339',
      'renewable': True,
      'title': 'Glass sword',
      'type': 'Bok',
      'year': '2015',
    }),
    dict({
      'author': 'Garber, Stephanie',
      'expire_date': '2025-03-19',
      'loan_date': '2025-02-19',
      'record_id': '488795',
      'renewable': True,
      'title': 'Finale',
      'type': 'Bok',
      'year': '2019',
    }),
    dict({
      'author': 'Garber, Stephanie',
      'expire_date': '2025-03-19',
      'loan_date': '2025-02-19',
      'record_id': '531601',
      'renewable': False,
      'title': 'Legendary',
      'type': 'Bok',
      'year': '2018',
    }),
  ])
# ---
# name: test_success_reservation_to_pick_up.1
  list([
    dict({
      'author': 'Yarros, Rebecca',
      'created_date': '2025-01-04',
      'pickup_date': '2025-02-26',
      'pickup_library': ' Stadsbiblioteket',
      'record_id': '641909',
      'reservation_number': '431',
      'title': 'Onyx storm',
      'type': 'Bok',
      'year': '2025',
    }),
  ])
# ---
